src/renderers/imgui_dx11/shader_program.rs

```diff
@@ -153,105 +162,124 @@ impl ShaderProgram {
-                StencilEnable: BOOL(0),
-                StencilReadMask: 0,
-                StencilWriteMask: 0,
-                FrontFace: D3D11_DEPTH_STENCILOP_DESC {
-                    StencilFailOp: D3D11_STENCIL_OP_KEEP,
-                    StencilDepthFailOp: D3D11_STENCIL_OP_KEEP,
-                    StencilPassOp: D3D11_STENCIL_OP_KEEP,
-                    StencilFunc: D3D11_COMPARISON_ALWAYS,
-                },
-                BackFace: D3D11_DEPTH_STENCILOP_DESC {
-                    StencilFailOp: D3D11_STENCIL_OP_KEEP,
-                    StencilDepthFailOp: D3D11_STENCIL_OP_KEEP,
-                    StencilPassOp: D3D11_STENCIL_OP_KEEP,
-                    StencilFunc: D3D11_COMPARISON_ALWAYS,
+        let mut depth_stencil_state = None;
+        unsafe {
+            dasc.dev().CreateDepthStencilState(
+                &D3D11_DEPTH_STENCIL_DESC {
+                    DepthEnable: BOOL(0),
+                    DepthFunc: D3D11_COMPARISON_ALWAYS,
+                    DepthWriteMask: D3D11_DEPTH_WRITE_MASK_ALL,
+                    StencilEnable: BOOL(0),
+                    StencilReadMask: 0,
+                    StencilWriteMask: 0,
+                    FrontFace: D3D11_DEPTH_STENCILOP_DESC {
+                        StencilFailOp: D3D11_STENCIL_OP_KEEP,
+                        StencilDepthFailOp: D3D11_STENCIL_OP_KEEP,
+                        StencilPassOp: D3D11_STENCIL_OP_KEEP,
+                        StencilFunc: D3D11_COMPARISON_ALWAYS,
+                    },
+                    BackFace: D3D11_DEPTH_STENCILOP_DESC {
+                        StencilFailOp: D3D11_STENCIL_OP_KEEP,
+                        StencilDepthFailOp: D3D11_STENCIL_OP_KEEP,
+                        StencilPassOp: D3D11_STENCIL_OP_KEEP,
+                        StencilFunc: D3D11_COMPARISON_ALWAYS,
+                    },
                 },
-            })?
+                Some(&mut depth_stencil_state),
+            )?
+        };
+
+        let shader_program = ShaderProgram {
+            vtx_shader: vtx_shader.unwrap(),
+            pix_shader: pix_shader.unwrap(),
+            layout: layout.unwrap(),
+            sampler: sampler.unwrap(),
+            blend_state: blend_state.unwrap(),
+            depth_stencil_state: depth_stencil_state.unwrap(),
+            rasterizer_state: rasterizer_state.unwrap(),
         };
 
-        Ok(ShaderProgram {
-            vtx_shader,
-            pix_shader,
-            layout,
-            sampler,
-            blend_state,
-            depth_stencil_state,
-            rasterizer_state,
-        })
+        Ok(shader_program)
     }
 
     pub(crate) unsafe fn set_state(&self, dasc: &DeviceAndSwapChain) {
-        dasc.dev_ctx().VSSetShader(&self.vtx_shader, &[]);
-        dasc.dev_ctx().PSSetShader(&self.pix_shader, &[]);
+        dasc.dev_ctx().VSSetShader(&self.vtx_shader, Some(&[]));
+        dasc.dev_ctx().PSSetShader(&self.pix_shader, Some(&[]));
```

Comment:
In this part specifically I have _no idea_ what passing in `Some(&[])` vs `None` does, and can't find any documentation.

Reply:
Me neither, I first tried passing just None & then changed it back to &[] because of the crashes, hoping that things would fall back in to place magically lol.

Reply:
From ChatGPT


VSSetShader is a function in DirectX used to set the vertex shader. It has two parameters:

1. pVertexShader: a pointer to the vertex shader. This parameter is required as it tells DirectX which vertex shader to use to process vertex data.

2. pClassInstances: a pointer to an array of D3D11_CLASS_INSTANCE structures. This parameter is optional and allows you to dynamically create vertex shader instances at runtime. If you don't need to create instances, you can set this parameter to NULL.

Other related parameters, such as shader constant buffers and input layout, need to be set before setting the vertex shader. The VSSetShader function simply binds the specified vertex shader to the vertex shader stage of the rendering pipeline.

It should be noted that this function is only applicable to DirectX 11 and higher versions. In earlier versions of DirectX, setting the vertex shader requires the use of different functions.

Reply:
Hehe, we're talking about the window-rs crate specifically. If only it was so easy to look at MDSN at all times..

Reply:
Hehe. I remember the parameter can be set to None